### PR TITLE
statistics: use top N of each store as hot regions

### PR DESCRIPTION
**Pull request by Luffbee**
### What problem does this PR solve? 

The original algorithm for finding hot regions cannot handle the "many warm region" case.

The "many warm region" case: 
Some stores are hot, because they have many "warm regions". Warm regions are regions that not hot enough to be regarded as hot regions.

### What is changed and how it works?

Regard the top N regions of each store as hot regions.

### Check List 

Tests 

 - Unit test

Side effects

 - Possible performance regression: 
  To maintain the top N regions, operations like `Put`, `Remove` will need at least `O(logN)`  time.

### CODE REVIEW

File: server/statistics/util.go [hunk 1]
```diff
@@ -53,3 +56,274 @@ func (r *RollingStats) Median() float64 {
...
+// NewTopN returns a TopN working on the items.
+func NewTopN(n int) *TopN {
+	return &TopN{
+		n:    maxInt(n, 1),
+		topn: newTopNHeap(),
+		rest: newRevTopNHeap(),
+	}
+}
+
+func (tn *TopN) lock() {
```
Comment (siddontang) [inline server/statistics/util.go]: you can use 

```go
type TopN struct {
    sync.RwMutex
}
```

so we don't need to provide these lock functions.

Comment (Luffbee) [reply server/statistics/util.go]: I don't want to export these lock functions.

Comment (siddontang) [reply server/statistics/util.go]: please remove these, too verbose. 

to follow the lock style in tidb, you can use 

```go
type TopN struct {
     mu        struct {
		sync.RWMutex
		// fields you want to protect
	}
}
```

Comment (Luffbee) [reply server/statistics/util.go]: Removed, but not the tidb lock style.

File: server/schedulers/hot_region.go [hunk 2]
```diff
@@ -220,12 +220,31 @@ func (h *balanceHotRegionsScheduler) balanceHotWriteRegions(cluster schedule.Clu
 	return nil
 }
 
-func calcScore(storeItems map[uint64][]*statistics.HotSpotPeerStat, cluster schedule.Cluster, kind core.ResourceKind) statistics.StoreHotRegionsStat {
+func calcScore(cluster schedule.Cluster, typ BalanceType, kind core.ResourceKind) statistics.StoreHotRegionsStat {
+	var storeItems map[uint64][]*statistics.HotSpotPeerStat
+	switch typ {
+	case hotWriteRegionBalance:
+		{
```
Comment (disksing) [inline server/schedulers/hot_region.go]: remove unnecessary parentheses?

File: server/statistics/hot_cache.go [hunk 3]
```diff
@@ -402,16 +407,30 @@ func (w *HotSpotCache) RandHotRegionFromStore(storeID uint64, kind FlowKind, hot
...
-		hotCacheStatusGauge.WithLabelValues("hotThreshold", storeTag, "write").Set(float64(threshold))
+		var topnMin, restMax uint64
+		if stat := flowStats.GetTopNMin(); stat != nil {
+			topnMin = stat.(*HotSpotPeerStat).GetFlowBytes()
+		}
+		if stat := flowStats.GetRestMax(); stat != nil {
+			restMax = stat.(*HotSpotPeerStat).GetFlowBytes()
+		}
+		hotCacheStatusGauge.WithLabelValues("topn_min", storeTag, "write").Set(float64(topnMin))
+		hotCacheStatusGauge.WithLabelValues("rest_max", storeTag, "write").Set(float64(restMax))
```
Comment (disksing) [inline server/statistics/hot_cache.go]: I guess that in most cases the two should be very close, and just recording one is enough.

**Comment by Luffbee** [comment]
I will close this PR, because it is part of the new hotspot scheduler, and we decided to work on the `new-hotspot-scheduler` branch now.

**Pull request closed without merging**